lib/active_merchant/billing/gateways/hi_pay.rb

```diff
@@ -0,0 +1,197 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class HiPayGateway < Gateway
+      self.test_url = 'https://stage-secure-gateway.hipay-tpp.com/rest'
+      self.live_url = 'https://secure-gateway.hipay-tpp.com/rest'
+
+      self.supported_countries = %w[FR]
+      self.default_currency = 'EUR'
+      self.money_format = :dollars
+      self.supported_cardtypes = %i[visa master american_express]
+
+      self.homepage_url = 'https://hipay.com/'
+      self.display_name = 'HiPay'
+
+      def initialize(options = {})
+        requires!(options, :username, :password)
+        @username = options[:username]
+        @password = options[:password]
+        super
+      end
+
+      def purchase(money, payment_method, options = {})
+        authorize(money, payment_method, options.merge({ operation: 'Sale' }))
+      end
+
+      def authorize(money, payment_method, options = {})
+        MultiResponse.run do |r|
+          if payment_method.is_a?(CreditCard)
+            response = r.process { tokenize(payment_method, options) }
+            card_token = response.params['token']
+          elsif payment_method.is_a?(String)
+            _transaction_ref, card_token, payment_product = payment_method.split('|')
+          end
+
+          post = {
+            payment_product: payment_product&.downcase || payment_method.brand,
+            operation: options[:operation] || 'Authorization',
+            cardtoken: card_token
+          }
+          add_address(post, options)
+          add_product_data(post, options)
+          add_invoice(post, money, options)
+          r.process { commit('order', post) }
+        end
+      end
+
+      def capture(money, authorization, options)
+        post = {}
+        post[:operation] = 'capture'
+        post[:currency] = (options[:currency] || currency(money))
+        transaction_ref, _card_token, _payment_product = authorization.split('|')
+        commit('capture', post, { transaction_reference: transaction_ref })
+      end
+
+      def store(payment_method, options = {})
+        tokenize(payment_method, options.merge({ multiuse: '1' }))
+      end
+
+      def scrub(transcrip)
+        # code
+      end
+
+      private
+
+      def add_product_data(post, options)
+        post[:orderid] = options[:order_id] if options[:order_id]
+        post[:description] = options[:description]
+      end
+
+      def add_invoice(post, money, options)
+        post[:currency] = (options[:currency] || currency(money))
+        post[:amount] = amount(money)
+      end
+
+      def add_credit_card(post, credit_card)
+        post[:card_number] = credit_card.number
+        post[:card_expiry_month] = credit_card.month
+        post[:card_expiry_year] = credit_card.year
+        post[:card_holder] = credit_card.name
+        post[:cvc] = credit_card.verification_value
+      end
+
+      def add_address(post, options)
+        if options[:billing_address]
```

Comment:
It might be nice to clean this method up a little by storing the `:billing_address` hash in a variable and then referencing the nested values like so:

```ruby
return unless billing_address = options[:billing_address]

post[:streetaddress] = billing_address[:address1]
...
```

Reply:
Thanks, yes that's nice.